File: main/cpp/system_info_lib.cpp
#include "system_info_lib.h"
#include <fstream>
#include <sstream>
#include <cstdlib>
#include <unistd.h>
#include <sys/utsname.h>
#include <dirent.h>

namespace SystemInfo {

SystemInfoLib::SystemInfoLib() {
}

SystemInfoLib::~SystemInfoLib() {
}

bool SystemInfoLib::read_dmi_file(const std::string& path, std::string& output) {
    std::ifstream file(path);
    if (!file.is_open()) {
        return false;
    }
    std::getline(file, output);
    file.close();
    return true;
}

std::string SystemInfoLib::read_file(const std::string& path, const std::string& default_value) {
    std::ifstream file(path);
    if (!file.is_open()) {
        return default_value;
    }
    std::string content;
    std::getline(file, content);
    file.close();
    return content;
}

bool SystemInfoLib::file_exists(const std::string& path) {
    std::ifstream file(path);
    return file.good();
}

HardwareInfo SystemInfoLib::get_hardware_info() {
    HardwareInfo info;
    
    info.system_vendor = read_file("/sys/class/dmi/id/sys_vendor", "N/A");
    info.product_name = read_file("/sys/class/dmi/id/product_name", "N/A");
    info.product_version = read_file("/sys/class/dmi/id/product_version", "N/A");
    info.board_vendor = read_file("/sys/class/dmi/id/board_vendor", "N/A");
    info.board_name = read_file("/sys/class/dmi/id/board_name", "N/A");
    info.bios_vendor = read_file("/sys/class/dmi/id/bios_vendor", "N/A");
    info.bios_version = read_file("/sys/class/dmi/id/bios_version", "N/A");
    info.bios_date = read_file("/sys/class/dmi/id/bios_date", "N/A");
    
    // CPU информация
    struct utsname uname_info;
    if (uname(&uname_info) == 0) {
        info.cpu_model = uname_info.machine;
    }
    
    // Память через /proc/meminfo
    std::ifstream meminfo("/proc/meminfo");
    if (meminfo.is_open()) {
        std::string line;
        while (std::getline(meminfo, line)) {
            if (line.find("MemTotal:") == 0) {
                std::istringstream iss(line);
                std::string label;
                long long mem_kb;
                iss >> label >> mem_kb;
                double mem_gb = mem_kb / (1024.0 * 1024.0);
                std::ostringstream oss;
                oss << mem_gb << " GB";
                info.total_memory = oss.str();
                break;
            }
        }
        meminfo.close();
    } else {
        info.total_memory = "N/A";
    }
    
    return info;
}

OSInfo SystemInfoLib::get_os_info() {
    OSInfo info;
    
    struct utsname uname_info;
    if (uname(&uname_info) == 0) {
        info.system = uname_info.sysname;
        info.release = uname_info.release;
        info.version = uname_info.version;
        info.machine = uname_info.machine;
        info.hostname = uname_info.nodename;
        info.kernel_version = uname_info.release;
    }
    
    // Дистрибутив из /etc/os-release
    std::ifstream os_release("/etc/os-release");
    if (os_release.is_open()) {
        std::string line;
        while (std::getline(os_release, line)) {
            if (line.find("PRETTY_NAME=") == 0) {
                size_t start = line.find('"');
                size_t end = line.rfind('"');
                if (start != std::string::npos && end != std::string::npos && start < end) {
                    info.distribution = line.substr(start + 1, end - start - 1);
                }
                break;
            }
        }
        os_release.close();
    } else {
        info.distribution = "N/A";
    }
    
    return info;
}

EnvironmentInfo SystemInfoLib::get_environment_info() {
    EnvironmentInfo info;
    
    // Десктопное окружение из переменных окружения
    const char* de = std::getenv("XDG_CURRENT_DESKTOP");
    info.desktop_environment = de ? de : "N/A";
    
    const char* ds = std::getenv("XDG_SESSION_TYPE");
    info.display_server = ds ? ds : "N/A";
    
    const char* shell = std::getenv("SHELL");
    info.shell = shell ? shell : "N/A";
    
    // Window manager через процессы
    if (info.desktop_environment == "N/A") {
        FILE* pipe = popen("ps -e", "r");
        if (pipe) {
            char buffer[128];
            std::string output;
            while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
                output += buffer;
            }
            pclose(pipe);
            
            std::string lower_output = output;
            for (auto& c : lower_output) {
                c = std::tolower(c);
            }
            
            if (lower_output.find("gnome") != std::string::npos) {
                info.desktop_environment = "GNOME";
            } else if (lower_output.find("kde") != std::string::npos) {
                info.desktop_environment = "KDE";
            } else if (lower_output.find("xfce") != std::string::npos) {
                info.desktop_environment = "XFCE";
            } else if (lower_output.find("mate") != std::string::npos) {
                info.desktop_environment = "MATE";
            }
        }
    }
    
    // Window manager
    FILE* wm_pipe = popen("ps aux | grep -i 'window_manager\\|i3\\|openbox\\|compiz'", "r");
    if (wm_pipe) {
        char buffer[128];
        std::string output;
        while (fgets(buffer, sizeof(buffer), wm_pipe) != nullptr) {
            output += buffer;
        }
        pclose(wm_pipe);
        
        if (output.find("i3") != std::string::npos) {
            info.window_manager = "i3";
        } else if (output.find("openbox") != std::string::npos) {
            info.window_manager = "Openbox";
        } else if (output.find("compiz") != std::string::npos) {
            info.window_manager = "Compiz";
        } else {
            info.window_manager = info.desktop_environment;
        }
    } else {
        info.window_manager = "N/A";
    }
    
    return info;
}

std::vector<BootloaderInfo> SystemInfoLib::get_bootloaders() {
    std::vector<BootloaderInfo> bootloaders;
    
    // Проверяем GRUB
    if (file_exists("/boot/grub/grub.cfg") || file_exists("/boot/grub2/grub.cfg")) {
        BootloaderInfo grub;
        grub.name = "GRUB";
        grub.path = file_exists("/boot/grub/grub.cfg") ? "/boot/grub/grub.cfg" : "/boot/grub2/grub.cfg";
        grub.active = true;
        bootloaders.push_back(grub);
    }
    
    // Проверяем systemd-boot
    if (file_exists("/boot/loader/loader.conf")) {
        BootloaderInfo systemd;
        systemd.name = "systemd-boot";
        systemd.path = "/boot/loader/loader.conf";
        systemd.active = true;
        bootloaders.push_back(systemd);
    }
    
    // Проверяем rEFInd
    if (file_exists("/boot/refind/refind.conf")) {
        BootloaderInfo refind;
        refind.name = "rEFInd";
        refind.path = "/boot/refind/refind.conf";
        refind.active = true;
        bootloaders.push_back(refind);
    }
    
    // Проверяем Windows Boot Manager через efibootmgr
    FILE* pipe = popen("efibootmgr 2>/dev/null", "r");
    if (pipe) {
        char buffer[512];
        std::string output;
        while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
            output += buffer;
        }
        pclose(pipe);
        
        if (output.find("Windows Boot Manager") != std::string::npos) {
            BootloaderInfo windows;
            windows.name = "Windows Boot Manager";
            windows.path = "EFI";
            windows.active = false;
            bootloaders.push_back(windows);
        }
    }
    
    return bootloaders;
}

std::vector<std::string> SystemInfoLib::get_installed_os() {
    std::vector<std::string> os_list;
    
    // Текущая ОС из /etc/os-release
    if (file_exists("/etc/os-release")) {
        std::ifstream os_release("/etc/os-release");
        std::string line;
        while (std::getline(os_release, line)) {
            if (line.find("PRETTY_NAME=") == 0) {
                size_t start = line.find('"');
                size_t end = line.rfind('"');
                if (start != std::string::npos && end != std::string::npos && start < end) {
                    os_list.push_back(line.substr(start + 1, end - start - 1) + " (Current)");
                }
                break;
            }
        }
        os_release.close();
    }
    
    // Проверяем GRUB конфигурацию
    if (file_exists("/boot/grub/grub.cfg")) {
        std::ifstream grub_cfg("/boot/grub/grub.cfg");
        std::string line;
        while (std::getline(grub_cfg, line)) {
            size_t pos = line.find("menuentry");
            if (pos != std::string::npos) {
                size_t start = line.find('"', pos);
                size_t end = line.find('"', start + 1);
                if (start != std::string::npos && end != std::string::npos) {
                    os_list.push_back(line.substr(start + 1, end - start - 1) + " (GRUB)");
                }
            }
        }
        grub_cfg.close();
    }
    
    // Проверяем Windows через /mnt
    DIR* dir = opendir("/mnt");
    if (dir) {
        struct dirent* entry;
        while ((entry = readdir(dir)) != nullptr) {
            std::string path = "/mnt/" + std::string(entry->d_name);
            if (file_exists(path + "/Windows") || file_exists(path + "/Program Files")) {
                os_list.push_back("Windows (Detected)");
                break;
            }
        }
        closedir(dir);
    }
    
    return os_list;
}

std::map<std::string, std::string> SystemInfoLib::get_bios_info() {
    std::map<std::string, std::string> bios;
    
    bios["vendor"] = read_file("/sys/class/dmi/id/bios_vendor", "N/A");
    bios["version"] = read_file("/sys/class/dmi/id/bios_version", "N/A");
    bios["release_date"] = read_file("/sys/class/dmi/id/bios_date", "N/A");
    bios["revision"] = read_file("/sys/class/dmi/id/bios_revision", "N/A");
    
    return bios;
}

std::vector<std::string> SystemInfoLib::get_drivers() {
    std::vector<std::string> drivers;
    
    // Получаем загруженные модули ядра
    FILE* pipe = popen("lsmod", "r");
    if (pipe) {
        char buffer[256];
        bool first_line = true;
        while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
            if (first_line) {
                first_line = false;
                continue; // Пропускаем заголовок
            }
            std::string line(buffer);
            size_t space_pos = line.find(' ');
            if (space_pos != std::string::npos) {
                drivers.push_back(line.substr(0, space_pos));
            }
        }
        pclose(pipe);
    }
    
    return drivers;
}

std::string SystemInfoLib::get_serial_number() {
    return read_file("/sys/class/dmi/id/product_serial", "N/A");
}

} // namespace SystemInfo

File: main/linux/system_info_lib.py
"""
Python обертка для C++ библиотеки системной информации
"""

import sys
import os

# Добавляем путь к собранной библиотеке
lib_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'build', 'lib')
if lib_path not in sys.path:
    sys.path.insert(0, lib_path)

try:
    # Пробуем импортировать C++ библиотеку
    import system_info_cpp
    CPP_AVAILABLE = True
except ImportError:
    CPP_AVAILABLE = False
    print("Warning: C++ system info library not available, using Python fallback")


class SystemInfo:
    """Обертка для системной информации"""
    
    def __init__(self):
        self.cpp_available = CPP_AVAILABLE
        if self.cpp_available:
            self.cpp_lib = system_info_cpp.SystemInfoLib()
    
    def get_hardware_info(self) -> dict:
        """Получить информацию о железе"""
        if self.cpp_available:
            hw = self.cpp_lib.get_hardware_info()
            return {
                'system_vendor': hw.system_vendor,
                'product_name': hw.product_name,
                'product_version': hw.product_version,
                'board_vendor': hw.board_vendor,
                'board_name': hw.board_name,
                'bios_vendor': hw.bios_vendor,
                'bios_version': hw.bios_version,
                'bios_date': hw.bios_date,
                'cpu_model': hw.cpu_model,
                'total_memory': hw.total_memory
            }
        else:
            return self._get_hardware_info_fallback()
    
    def get_os_info(self) -> dict:
        """Получить информацию об ОС"""
        if self.cpp_available:
            os_info = self.cpp_lib.get_os_info()
            return {
                'system': os_info.system,
                'release': os_info.release,
                'version': os_info.version,
                'machine': os_info.machine,
                'hostname': os_info.hostname,
                'distribution': os_info.distribution,
                'kernel_version': os_info.kernel_version
            }
        else:
            return self._get_os_info_fallback()
    
    def get_environment_info(self) -> dict:
        """Получить информацию об окружении"""
        if self.cpp_available:
            env = self.cpp_lib.get_environment_info()
            return {
                'desktop_environment': env.desktop_environment,
                'display_server': env.display_server,
                'shell': env.shell,
                'window_manager': env.window_manager
            }
        else:
            return self._get_environment_info_fallback()
    
    def get_bootloaders(self) -> list:
        """Получить список загрузчиков"""
        if self.cpp_available:
            bootloaders = self.cpp_lib.get_bootloaders()
            return [
                {
                    'name': bl.name,
                    'path': bl.path,
                    'active': bl.active
                }
                for bl in bootloaders
            ]
        else:
            return self._get_bootloaders_fallback()
    
    def get_installed_os(self) -> list:
        """Получить список установленных ОС"""
        if self.cpp_available:
            return self.cpp_lib.get_installed_os()
        else:
            return self._get_installed_os_fallback()
    
    def get_bios_info(self) -> dict:
        """Получить информацию о BIOS"""
        if self.cpp_available:
            return self.cpp_lib.get_bios_info()
        else:
            return self._get_bios_info_fallback()
    
    def get_drivers(self) -> list:
        """Получить список драйверов"""
        if self.cpp_available:
            return self.cpp_lib.get_drivers()
        else:
            return self._get_drivers_fallback()
    
    def get_serial_number(self) -> str:
        """Получить серийный номер"""
        if self.cpp_available:
            return self.cpp_lib.get_serial_number()
        else:
            return self._get_serial_number_fallback()
    
    def _get_hardware_info_fallback(self) -> dict:
        """Python fallback для информации о железе"""
        import platform
        info = {
            'system_vendor': 'N/A',
            'product_name': 'N/A',
            'product_version': 'N/A',
            'board_vendor': 'N/A',
            'board_name': 'N/A',
            'bios_vendor': 'N/A',
            'bios_version': 'N/A',
            'bios_date': 'N/A',
            'cpu_model': platform.processor(),
            'total_memory': 'N/A'
        }
        
        # Читаем DMI файлы
        dmi_files = {
            'system_vendor': '/sys/class/dmi/id/sys_vendor',
            'product_name': '/sys/class/dmi/id/product_name',
            'product_version': '/sys/class/dmi/id/product_version',
            'board_vendor': '/sys/class/dmi/id/board_vendor',
            'board_name': '/sys/class/dmi/id/board_name',
            'bios_vendor': '/sys/class/dmi/id/bios_vendor',
            'bios_version': '/sys/class/dmi/id/bios_version',
            'bios_date': '/sys/class/dmi/id/bios_date'
        }
        
        for key, path in dmi_files.items():
            try:
                with open(path, 'r') as f:
                    info[key] = f.read().strip()
            except (FileNotFoundError, IOError, PermissionError):
                pass
        
        try:
            import psutil
            mem = psutil.virtual_memory()
            info['total_memory'] = f"{mem.total / (1024**3):.2f} GB"
        except ImportError:
            pass
        
        return info
    
    def _get_os_info_fallback(self) -> dict:
        """Python fallback для информации об ОС"""
        import platform
        info = {
            'system': platform.system(),
            'release': platform.release(),
            'version': platform.version(),
            'machine': platform.machine(),
            'hostname': platform.node(),
            'distribution': 'N/A',
            'kernel_version': platform.release()
        }
        
        # Читаем /etc/os-release
        try:
            with open('/etc/os-release', 'r') as f:
                for line in f:
                    if line.startswith('PRETTY_NAME='):
                        name = line.split('=', 1)[1].strip('"')
                        info['distribution'] = name
                        break
                    elif line.startswith('NAME='):
                        name = line.split('=', 1)[1].strip('"')
                        if info['distribution'] == 'N/A':
                            info['distribution'] = name
        except (FileNotFoundError, IOError):
            pass
        
        return info
    
    def _get_environment_info_fallback(self) -> dict:
        """Python fallback для информации об окружении"""
        import os
        return {
            'desktop_environment': os.environ.get('XDG_CURRENT_DESKTOP', 'N/A'),
            'display_server': os.environ.get('XDG_SESSION_TYPE', 'N/A'),
            'shell': os.environ.get('SHELL', 'N/A'),
            'window_manager': 'N/A'
        }
    
    def _get_bootloaders_fallback(self) -> list:
        """Python fallback для загрузчиков"""
        bootloaders = []
        import os
        
        if os.path.exists('/boot/grub/grub.cfg'):
            bootloaders.append({'name': 'GRUB', 'path': '/boot/grub/grub.cfg', 'active': True})
        
        if os.path.exists('/boot/loader/loader.conf'):
            bootloaders.append({'name': 'systemd-boot', 'path': '/boot/loader/loader.conf', 'active': True})
        
        return bootloaders
    
    def _get_installed_os_fallback(self) -> list:
        """Python fallback для установленных ОС"""
        os_list = []
        import os
        
        try:
            with open('/etc/os-release', 'r') as f:
                for line in f:
                    if line.startswith('PRETTY_NAME='):
                        name = line.split('=', 1)[1].strip('"')
                        os_list.append(f"{name} (Current)")
                        break
        except (FileNotFoundError, IOError):
            pass
        
        return os_list
    
    def _get_bios_info_fallback(self) -> dict:
        """Python fallback для BIOS"""
        bios = {}
        
        # Читаем DMI файлы BIOS
        dmi_files = {
            'vendor': '/sys/class/dmi/id/bios_vendor',
            'version': '/sys/class/dmi/id/bios_version',
            'release_date': '/sys/class/dmi/id/bios_date',
            'revision': '/sys/class/dmi/id/bios_revision'
        }
        
        for key, path in dmi_files.items():
            try:
                with open(path, 'r') as f:
                    bios[key] = f.read().strip()
            except (FileNotFoundError, IOError, PermissionError):
                bios[key] = 'N/A'
        
        return bios
    
    def _get_drivers_fallback(self) -> list:
        """Python fallback для драйверов"""
        drivers = []
        import subprocess
        
        try:
            result = subprocess.run(['lsmod'], capture_output=True, text=True, timeout=5)
            lines = result.stdout.split('\n')[1:]  # Пропускаем заголовок
            for line in lines:
                parts = line.split()
                if parts:
                    drivers.append(parts[0])
        except (subprocess.TimeoutExpired, FileNotFoundError):
            pass
        
        return drivers[:50]  # Ограничиваем первые 50 драйверов
    
    def _get_serial_number_fallback(self) -> str:
        """Python fallback для серийного номера"""
        try:
            with open('/sys/class/dmi/id/product_serial', 'r') as f:
                return f.read().strip()
        except (FileNotFoundError, IOError, PermissionError):
            return 'Требуются root права'
